engine/common/grpc/forwarder/forwarder.go

```diff
@@ -62,57 +58,24 @@ func (f *Forwarder) setFlowAccessAPI(accessNodeAddressAndPort flow.IdentityList,
 	return nil
 }
 
-// reconnectingClient returns an active client, or
-// creates one, if the last one is not ready anymore.
+// reconnectingClient returns an active client, or creates a new connection.
 func (f *Forwarder) reconnectingClient(i int) error {
-	timeout := f.timeout
-
-	if f.connections[i] == nil || f.connections[i].GetState() != connectivity.Ready {
-		identity := f.ids[i]
-		var connection *grpc.ClientConn
-		var err error
-		if identity.NetworkPubKey == nil {
-			connection, err = grpc.Dial(
-				identity.Address,
-				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(int(f.maxMsgSize))),
-				grpc.WithTransportCredentials(insecure.NewCredentials()),
-				rpcConnection.WithClientTimeoutOption(timeout))
-			if err != nil {
-				return err
-			}
-		} else {
-			tlsConfig, err := grpcutils.DefaultClientTLSConfig(identity.NetworkPubKey)
-			if err != nil {
-				return fmt.Errorf("failed to get default TLS client config using public flow networking key %s %w", identity.NetworkPubKey.String(), err)
-			}
-
-			connection, err = grpc.Dial(
-				identity.Address,
-				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(int(f.maxMsgSize))),
-				grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
-				rpcConnection.WithClientTimeoutOption(timeout))
-			if err != nil {
-				return fmt.Errorf("cannot connect to %s %w", identity.Address, err)
-			}
-		}
-		connection.Connect()
-		time.Sleep(1 * time.Second)
-		state := connection.GetState()
-		if state != connectivity.Ready && state != connectivity.Connecting {
-			return fmt.Errorf("%v", state)
-		}
-		f.connections[i] = connection
-		f.upstream[i] = access.NewAccessAPIClient(connection)
-	}
+	identity := f.upstream[i].id
 
+	accessApiClient, closer, err := f.connFactory.GetAccessAPIClientWithPort(identity.Address, identity.NetworkPubKey)
+	if err != nil {
+		return fmt.Errorf("failed to connect to access node at %s: %w", accessApiClient, err)
+	}
+	f.upstream[i].closer = closer
+	f.upstream[i].client = accessApiClient
 	return nil
 }
 
 // FaultTolerantClient implements an upstream connection that reconnects on errors
 // a reasonable amount of time.
```

Comment:
I think we should describe in documentation that closer is not nil iff err is nil, otherwise it's not clear when closer should be closed